homeassistant/components/decora_wifi/light.py

```diff
@@ -2,90 +2,52 @@
-import homeassistant.helpers.config_validation as cv
-
-_LOGGER = logging.getLogger(__name__)
 
-# Validation of the user's configuration
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {vol.Required(CONF_USERNAME): cv.string, vol.Required(CONF_PASSWORD): cv.string}
+from .common import (
+    DecoraWifiCommFailed,
+    DecoraWifiEntity,
+    DecoraWifiPlatform,
+    DecoraWifiSessionNotFound,
 )
+from .const import CONF_LIGHT, DOMAIN
 
-NOTIFICATION_ID = "leviton_notification"
-NOTIFICATION_TITLE = "myLeviton Decora Setup"
+_LOGGER = logging.getLogger(__name__)
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up the Decora WiFi platform."""
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Decora Wifi lights based on a config entry."""
 
-    email = config[CONF_USERNAME]
-    password = config[CONF_PASSWORD]
-    session = DecoraWiFiSession()
+    email = hass.data[DOMAIN][entry.entry_id]
 
     try:
-        success = session.login(email, password)
-
-        # If login failed, notify user.
-        if success is None:
-            msg = "Failed to log into myLeviton Services. Check credentials."
-            _LOGGER.error(msg)
-            hass.components.persistent_notification.create(
-                msg, title=NOTIFICATION_TITLE, notification_id=NOTIFICATION_ID
-            )
-            return False
-
-        # Gather all the available devices...
-        perms = session.user.get_residential_permissions()
-        all_switches = []
-        for permission in perms:
-            if permission.residentialAccountId is not None:
-                acct = ResidentialAccount(session, permission.residentialAccountId)
-                for residence in acct.get_residences():
-                    for switch in residence.get_iot_switches():
-                        all_switches.append(switch)
-            elif permission.residenceId is not None:
-                residence = Residence(session, permission.residenceId)
-                for switch in residence.get_iot_switches():
-                    all_switches.append(switch)
-
-        add_entities(DecoraWifiLight(sw) for sw in all_switches)
-    except ValueError:
-        _LOGGER.error("Failed to communicate with myLeviton Service")
-
-    # Listen for the stop event and log out.
-    def logout(event):
-        """Log out..."""
-        try:
-            if session is not None:
-                Person.logout(session)
-        except ValueError:
-            _LOGGER.error("Failed to log out of myLeviton Service")
-
-    hass.bus.listen(EVENT_HOMEASSISTANT_STOP, logout)
-
-
-class DecoraWifiLight(LightEntity):
+        devices = await DecoraWifiPlatform.async_getdevices(hass, email)
+    except DecoraWifiCommFailed:
+        _LOGGER.error("Communication with Decora Wifi platform failed.")
+    except DecoraWifiSessionNotFound:
+        _LOGGER.error("DecoraWifi Session not found.")
+    finally:
```

Comment:
why use finally here? If there are no devices, there are no entities to be added, so you could just return?

Reply:
Agreed, it's unnecessary here.